src/x.c

```diff
@@ -15,91 +15,148 @@ int16_t x_parseFile(
                     routesCalled.buffer[i] ++;
                 }
+
             }
+
             totalMatched ++;
         }
-        totalLines ++;
 
+        totalLines ++;
         bytesRead += strlen(line) + 1;
-
         if (!(lineCount % 5432)) {
             for (i = 0; i < 80; i++) {
                 printf("\b");
             }
-            printf("x/parser: %.1f%% (total = %d, matched = %d)", 
+
+            printf("x/parser: %.1f%% (total = %d, matched = %d)",
                 100.0 * ((float)bytesRead / (float)size), totalLines, totalMatched);
             fflush(stdout);
         }
+
         lineCount ++;
     }
+
     corto_timeGet(&stop);
     for (i = 0; i < 80; i++) {
         printf("\b");
     }
-    printf("x/parser: 100%% (total = %d, matched = %d)  \n", 
+
+    printf("x/parser: 100%% (total = %d, matched = %d)  \n",
         totalLines, totalMatched);
     printf("x/parser: done (%.1f seconds)\n",
         corto_timeToDouble(corto_timeSub(stop, start)));
-
     corto_trace("");
     corto_trace("x: Summary:");
     for (i = 0; i < routesCalled.length; i++) {
-        corto_trace("x: %30s: called %d times", 
-            corto_idof(methods->buffer[i]), 
+        corto_trace("x: %30s: called %d times",
+            corto_idof(methods->buffer[i]),
             routesCalled.buffer[i]);
     }
+
     corto_trace("");
-    corto_trace("x: parsed %d lines, matched %d in %fs", 
-        totalLines, 
+    corto_trace("x: parsed %d lines, matched %d in %fs",
+        totalLines,
         totalMatched,
         corto_timeToDouble(corto_timeSub(stop, start)));
-
     corto_dealloc(routesCalled.buffer);
 
+    fclose(f); /* Close file handler */
+
     return 0;
 error:
     return -1;
 }
 
+int16_t x_parseString(
+    corto_object instance,
+    corto_string line,
+    corto_object visitor)
+{
+    corto_routerimpl parser = corto_routerimpl(corto_typeof(instance));
+    corto_objectseq *methods = &corto_interface(parser)->methods;
+    corto_int32seq routesCalled;
+    routesCalled.buffer = corto_calloc(methods->length * sizeof(corto_int32));
+    routesCalled.length = methods->length;
+    corto_int32 matched = 0;
+    corto_time start, stop;
+    corto_route route = NULL;
+    corto_int32 i;
+    corto_any param = {NULL}, result = {NULL, NULL};
+    if (visitor) {
+        param.type = corto_typeof(visitor);
+        param.value = visitor;
+    }
+
+    /* Benchmarking only enabled when Corto running with TRACE verbosity */
+    if (corto_verbosityGet() <= CORTO_TRACE) {
+        corto_trace("x/parser: parsing [%s]", line);
+        corto_timeGet(&start);
+    }
+
+    if (corto_router_match(instance, line, param, result, &route)) {
+        corto_lasterr(); /* Suppress uncatched error warning */
+    } else if (route) {
+        /* Keep track of how many times a route is called */
+        for (i = 0; i < methods->length; i++) {
+            if (methods->buffer[i] == route) {
+                routesCalled.buffer[i] ++;
```

Comment:
This code is not necessary. Since if you are only parsing a single line, you'll only match a single route, which is stored in the `route` variable.

Reply:
@SanderMertens done.